### PR TITLE
Change Frame.texture_coordinates() to return [f32; 2]

### PR DESCRIPTION
**Pull request by sunsided**
This pull request is related to issue #14 and changes the `texture_coordinates()` method of the `Frame` struct; it now returns instances of type `texture_coordinate` instead of `rs2_point` values.

In the `librealsense` C API, `rs2_get_frame_texture_coordinates` is defined [here](https://github.com/IntelRealSense/librealsense/blob/0adceb9dc6fce63c348346e1aef1b63c052a1db9/include/librealsense2/h/rs_frame.h#L217-L224) as

```c
rs2_pixel* rs2_get_frame_texture_coordinates(const rs2_frame* frame, rs2_error** error);
```

with `rs2_pixel` being defined [here](https://github.com/IntelRealSense/librealsense/blob/0adceb9dc6fce63c348346e1aef1b63c052a1db9/include/librealsense2/h/rs_types.h#L116-L120) as

```c
typedef struct rs2_pixel
{
    int ij[2];
} rs2_pixel;
```

that is, as a struct containing an `[i32; 2]`; bindgen is correctly creating a wrapper for that. However, the C++ API defines `get_texture_coordinates()` [here](https://github.com/IntelRealSense/librealsense/blob/0adceb9dc6fce63c348346e1aef1b63c052a1db9/include/librealsense2/hpp/rs_frame.hpp#L765-L775) as

```cpp
const texture_coordinate* get_texture_coordinates() const
{
    rs2_error* e = nullptr;
    auto res = rs2_get_frame_texture_coordinates(get(), &e);
    error::handle(e);
    return (const texture_coordinate*)res;
}
```

where `texture_coordinate` is

```cpp
struct texture_cstruct texture_coordinate {
    float u, v;
    operator const float*() const { return &u; }
};
```

In other words, the underlying texture coordinates are actually `f32` instead of `i32` and callers of the Rust method would need to transmute the coordinates first to make sense of them.

I added this functionality to the Rust API by introducing the type

```rust
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct texture_coordinate {
    pub u: f32,
    pub v: f32,
}
```

and returning that instead. The examples are updated accordingly and now render colored point clouds.

### CODE REVIEW

**Comment by jerry73204** [comment]
Yes, the FFI type is currently a issue and thanks that you pointed it out. Before that, we have to be cautious on memory layout of data. Your change involves in reinterpreting the data by pointer casting. I need a confidence that the compiler makes a guarantee on the memory layout is also valid.

File: src/frame.rs [hunk 1]
```diff
@@ -756,6 +756,14 @@ impl Frame<marker::Pose> {
     }
 }
 
+/// UV texture coordinates.
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct texture_coordinate {
```
Comment (jerry73204) [inline src/frame.rs]: Please rename to `TextureCoordinate` to follow the Rust's naming convention.

File: src/frame.rs [hunk 2]
```diff
@@ -771,16 +779,16 @@ impl Frame<marker::Points> {
...
-    pub fn texture_coordinates<'a>(&'a self) -> RsResult<&'a [realsense_sys::rs2_pixel]> {
+    pub fn texture_coordinates<'a>(&'a self) -> RsResult<&'a [texture_coordinate]> {
         unsafe {
             let n_points = self.points_count()?;
             let mut checker = ErrorChecker::new();
             let ptr = realsense_sys::rs2_get_frame_texture_coordinates(
                 self.ptr.as_ptr(),
                 checker.inner_mut_ptr(),
-            );
+            ) as *const texture_coordinate;
```
Comment (jerry73204) [inline src/frame.rs]: Please give a reference that the casting is safe, or use `safe_transmute` instead.

**Review by jerry73204 — COMMENTED** [review]
Please see the review inline. I'll merge the PR soon once it's done.

**Comment by sunsided** [comment]
Sorry - I'm somewhat new to Rust, so that's possibly a bit off topic - could you provide me a pointer to why the cast could be unsafe? 

I might be missing something, but since the C++ code [here](https://github.com/IntelRealSense/librealsense/blob/0adceb9dc6fce63c348346e1aef1b63c052a1db9/include/librealsense2/hpp/rs_frame.hpp#L765-L775) only casts the `rs2_point*` value returned by the C API, the memory layout between the two of them needs to be identical. Since bindgen already directly maps the C type to a Rust type, shouldn't the other cast succeed as well (`#[repr(C)]` and all)?

**Comment by jerry73204** [comment]
I see the memory layout is undefined by default due compiler optimization ([ref](https://doc.rust-lang.org/reference/types/struct.html)). That's the way C differs from Rust. I just read the [doc](https://doc.rust-lang.org/reference/type-layout.html#the-c-representation) and gained some confidence your case is fine with  `repr(C)`.

**Comment by sunsided** [comment]
I just realized that the underlying assumption `::std::os::raw::c_int` == `i32` is not generally valid. Interestingly, according to [this post](https://stackoverflow.com/a/11833229/195651) even the librealsense2 assumption that `sizeof(int) == sizeof(float)` would be broken. Is there a smart way to get a floating-point type of the same size as `::std::os::raw::c_int`, regardless of the system?

**Comment by jerry73204** [comment]
Yes, that's why I consider the `safe_transmute`. At least this point could be checked in compile time.

**Comment by sunsided** [comment]
Cool. `safe_transmute` is new to me, but I'll try to wrap my head around that one. 🙂:v:

File: src/frame.rs (hunk 2 above)
Comment (sunsided) [reply src/frame.rs]: Added a safety comment and used `safe_transmute::transmute_many()` to convert from `*const rs2_pixel` to `&[TextureCoordinate]`.

**Comment by sunsided** [comment]
I have a vague feeling that I'm using `transmute_many()` wrong, but the way it's implemented now is: `*const rs2_pixel` → `&[rs2_pixel]` → `*const u8` -> `&[u8]` → `&[TextureCoordinate]`.

The detour over `&[rs2_pixel]` is a bit much, but this way it provides some symmetry with the implementation of `vertices()` and allows to make use of `std::mem::size_of_val(slice)` rather than using `n_points * std::mem::size_of::<rs2_pixel>()`. I can remove that one though, this way it'd end up at two lines and an assertion.

`TextureCoordinate` needs to implement `TriviallyTransmutable` for `transmute_many` to work, but since it's a POD type it should be fine if I got the docs right.

**Comment by jerry73204** [comment]
I think `repr[C]` gives us a guarantee. In fact, I learned a trick from [this post](https://blog.mozilla.org/nnethercote/2018/11/09/how-to-get-the-size-of-rust-types-with-zprint-type-sizes/) and checked it accordingly. The impl looks fine. Thanks for your work.
